dunereco/FDSensOpt/NeutrinoAngularRecoAlg/NeutrinoAngularRecoAlg.cc

```diff
@@ -323,4 +520,81 @@ bool NeutrinoAngularRecoAlg::IsTrackContained(const art::Ptr<recob::Track> &pTra
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+
+CalorimetricDirectionFitter::CalorimetricDirectionFitter(std::vector<PolarFitOutput>& observed_data, double thetaU, double thetaV) :
+    _thetaU(thetaU), _thetaV(thetaV)
+{
+
+    for(const PolarFitOutput& view_fit : observed_data){
+        if(!view_fit.success){
+            continue;
+        }
+
+        switch (view_fit.view)
+        {
+            case geo::View_t::kU:
+                _pU = view_fit.pview;
+                _pxU = view_fit.pdrift;
+                _nhitsU = view_fit.nhits;
+                break;
+            case geo::View_t::kV:
+                _pV = view_fit.pview;
+                _pxV = view_fit.pdrift;
+                _nhitsV = view_fit.nhits;
+                break;
+            case geo::View_t::kW:
+                _pW = view_fit.pview;
+                _pxW = view_fit.pdrift;
+                _nhitsW = view_fit.nhits;
+                break;
+            default:
+                break;
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+
+double CalorimetricDirectionFitter::Chi2(double const* x) const{
+    double theta = x[0];
+    double phi = x[1];
+
+    double px = sin(theta)*cos(phi);
+    double py = sin(theta)*sin(phi);
+    double pz = cos(theta);
+
+    double pu = cos(_thetaU)*pz - sin(_thetaU)*py;
+    double pv = cos(_thetaV)*pz - sin(_thetaV)*py;
+
+    double u_theta_meas = atan2(_pU, _pxU);
+    double v_theta_meas = atan2(_pV, _pxV);
+    double w_theta_meas = atan2(_pW, _pxW);
+
+    double u_theta_fit = atan2(pu, px);
+    double v_theta_fit = atan2(pv, px);
+    double w_theta_fit = atan2(pz, px);
+
+    double chi2 = 0;
+    if (_nhitsU > 0){
+        chi2 += (u_theta_fit - u_theta_meas)*(u_theta_fit - u_theta_meas);
+    }
+    if(_nhitsV > 0){
+        chi2 += (v_theta_fit - v_theta_meas)*(v_theta_fit - v_theta_meas);
+    }
+    if(_nhitsW > 0)
+    {
+        chi2 += (w_theta_fit - w_theta_meas)*(w_theta_fit - w_theta_meas);
+    }
+
+    return chi2;
```

Comment:
I don't think this particularly matters for the fit output, but it would be worth considering including an uncertainty here, and normalising by the number of degrees of freedom, as someone down the line might try to interpret this chi2 value.